trinity/plugins/builtin/ethstats/plugin.py

```diff
@@ -0,0 +1,63 @@
+import asyncio
+
+from argparse import (
+    ArgumentParser,
+    _SubParsersAction,
+)
+
+from trinity.extensibility import (
+    BaseEvent,
+    BasePlugin,
+    PluginContext,
+)
+from trinity.extensibility.events import (
+    ResourceAvailableEvent,
+    TrinityStartupEvent,
+)
+from eth.chains.base import (
+    BaseChain
+)
+from p2p.peer import (
+    PeerPool
+)
+
+from trinity.plugins.builtin.ethstats.ethstats_service import (
+    EthstatsService,
+)
+
+
+class EthstatsPlugin(BasePlugin):
+
+    def __init__(self) -> None:
+        self.is_enabled: bool = False
+
+        self.chain: BaseChain = None
+        self.peer_pool: PeerPool = None
+
+    @property
+    def name(self) -> str:
+        return "Ethstats"
+
+    def configure_parser(self, arg_parser: ArgumentParser, subparser: _SubParsersAction) -> None:
+        arg_parser.add_argument(
+            "--ethstats",
+            action="store_true",
+            help="Enable node status reporting service (experimental)",
+        )
+
+    def handle_event(self, activation_event: BaseEvent) -> None:
+        if isinstance(activation_event, TrinityStartupEvent):
+            self.is_enabled = activation_event.args.ethstats
+        if isinstance(activation_event, ResourceAvailableEvent):
+            if activation_event.resource_type is PeerPool:
+                self.peer_pool = activation_event.resource
+            elif activation_event.resource_type is BaseChain:
+                self.chain = activation_event.resource
+
+    def should_start(self) -> bool:
+        return all((self.is_enabled, self.chain is not None, self.peer_pool is not None))
+
+    def start(self) -> None:
+        service = EthstatsService('ws://localhost:3000/api', 'SECRET', self.chain, self.peer_pool)
+
+        asyncio.ensure_future(service.run())
```

Comment:
Plugins that run in the networking process do currently not receive any notification about cancellation. In the case of e.g. the `TxPool` plugin that isn't a real issue as that one chains with the `CancelToken` of the node and hence at least gets shut down cleanly with the node.

https://github.com/ethereum/py-evm/blob/d225f4a26811aa15f450378aa8d618caa573b158/trinity/plugins/builtin/tx_pool/plugin.py#L84

[There's a PR coming](https://github.com/ethereum/py-evm/pull/1309) that will guarantee a cleaner teardown story for networking plugins, however, if this would move into a `BaseIsolatedPlugin` the teardown story would change a bit as these plugins run in their own process and receive `SIGINT` / `SIGTERM` that need to be handled.

https://github.com/ethereum/py-evm/blob/d225f4a26811aa15f450378aa8d618caa573b158/trinity/plugins/builtin/json_rpc/plugin.py#L67

Reply:
Refactored plugin to be `BaseIsolatedPlugin`.